[WIP] v2

steelbrain opened the pull request:
- [x] `pundle-resolver-default`
- [x] `pundle-loader-json`
- [x] `pundle-loader-js`
- [ ] ~~`pundle-loader-file`~~
- [x] `pundle-loader-url`
- [x] `pundle-transformer-babel`
- [x] `pundle-post-transformer-uglify`
- [x] `pundle-plugin-npm-installer`
- [x] `pundle-preset-default`
- [x] `pundle-generator-default`
- [x] `pundle-reporter-cli`
- [ ] ~~`pundle-plugin-eslint`~~
- [x] `pundle-cli`
- [x] `pundle-dev`
- [x] `pundle`
  - [ ] ~~code splitting~~

steelbrain committed:
:fire: Remove default loaders and cli from main package

steelbrain committed:
:new: Update signature of pundle-api callbacks to allow source

steelbrain committed:
:fire: Cleanup main Pundle package

steelbrain committed:
:new: Add publicPath to pundle api File spec

steelbrain committed:
:new: Validate and fill new config schema properly

steelbrain committed:
:fire: Cleanup pundle manifest

steelbrain committed:
:new: Implement Component loader in Pundle

steelbrain committed:
:new: Add a dummy loader-js package

steelbrain committed:
:new: Add a working loader-json package

steelbrain committed:
:art: Minor tweak to ReporterCallback signature

steelbrain committed:
:new: Implement Babel Transformer using pundle-api

steelbrain committed:
:new: Add spec dirs to packages

steelbrain committed:
:fire: Nuke generator and dev as outdated

steelbrain committed:
:new: Add resolver-default and cleanup some code

steelbrain committed:
:bug: Strip-bom in pundle-fs

steelbrain committed:
:new: Implement pundle-resolver-default 🎉
This has got to be our best resolver to date, fully compliant with the browser spec and more configurable than ever

steelbrain committed:
:new: Implement pundle-plugin-npm-installer

steelbrain committed:
:art: Fix code duplication in pundle-resolver-default

steelbrain committed:
:new: Add pundle-post-transformer-uglify

steelbrain committed:
:fire: Remove the Files class from Pundle
The API I have in mind right now requires us to keep the file cache in each consumer of the Pundle API rather than in a class, so the consumer has full control over how it's used and the lifecycle is simpler and more pipeline type.
After this class is removed, Pundle will not persist any file related thing in itself (individual components still can, and should, unless asked otherwise). Pundle will have a function where you pass the file path, it'll take it, load it, transform it, execute the plugins on it and return you it's sources, contents, sourceMap and the imports.
Then the consumer and use a do while loop to keep resolving the requires (keeping in mind dupes) until he has all.

steelbrain committed:
:new: Implement Emitter in pundle-api

steelbrain committed:
:bug: Minor bugfixes and API changes

steelbrain committed:
:art: Change API signature a bit

steelbrain committed:
:new: Add Compilation class to pundle

steelbrain committed:
:new: Add allowRecursive to createResolver() options

steelbrain committed:
:new: Implement Compilation::resolve

steelbrain committed:
:new: Implement file processing in Pundle's compilation

steelbrain committed:
:new: Merge source maps between steps properly

steelbrain committed:
:bug: Fix CI

steelbrain committed:
:bug: Misc bugfixes

steelbrain committed:
:new: Add preset support to Pundle

steelbrain committed:
:new: Add pundle-preset-default

steelbrain committed:
:new: Usability fixes in preset API

steelbrain committed:
:new: Execute Transformers *before* Loaders
The reason being that Transformers might want to handle some crazy syntax before loaders can work on it

steelbrain committed:
:new: Implement replacement/requires support in pundle-loader-js

steelbrain committed:
:new: Use require instead of __require

steelbrain committed:
:new: Implement processTree in Compilation and Pundle

steelbrain committed:
:new: Implement generate() in Compilatio and Pundle

steelbrain committed:
:memo: Write specs of loadPreset()

steelbrain committed:
:art: Make generator output more flexible

steelbrain committed:
:new: Add struct for generator-default

steelbrain committed:
:new: Make generator-default and add wrappers

appsforartists commented:
Curious how far out you think this is

steelbrain commented:
It's actually mostly done

Here's what it can do now
- It can generate bundles for different file types while complying a spec
- It can load presets and plugins

Here's what still needs a little love
- Dev Server
- Watching and recompiling

appsforartists commented:
The dev server is the bit I'm interested in, so I'll hold out til it's ready.  😀

steelbrain committed:
:art: Minor fixes for generator-default

steelbrain committed:
:new: Add wrapper type "none"

steelbrain committed:
:new: Standardize imports

steelbrain committed:
:new: Make generator work completely

steelbrain committed:
:new: Make normal wrapper work

steelbrain committed:
:new: Add a mock wrapper as HMR (copied from normal)

steelbrain committed:
:new: Add activate() and dispose() support to all components

steelbrain committed:
:minibus: publicDirectory -> publicPath

steelbrain committed:
:new: Implement watcher in Pundle

steelbrain committed:
:new: Invoke PostTransformers once instead of per file
It's much better to compile the bundle at once, entirely with uglify than one by one, leaving out wrapper and it's vars

steelbrain committed:
:new: Minify wrappers and allow non-strict-mode

steelbrain committed:
:new: Re-arrange tree processing functions and add specs

steelbrain committed:
:new: Add pundle-reporter-cli package

steelbrain committed:
:new: Use new error classes throughout Pundle

steelbrain committed:
🚐 {File,Message}Error -> {File,Message}Issue

steelbrain committed:
:new: Use relative file path in error messages

steelbrain committed:
:new: Add support for source map path and inline in generator

steelbrain committed:
:bug: Fix offsets in source maps

steelbrain committed:
:new: Add fake data in json source map to make it show up in dev tools

steelbrain committed:
:new: Use compact instead of concise

steelbrain committed:
:new: Lay foundations for Pundleconfig.js

steelbrain committed:
:art: Fix code duplication and fragmentation

steelbrain committed:
:tada: Implement first-class support for Pundleconfig.js

steelbrain committed:
:new: Add skeleton files for Pundle-Dev

steelbrain committed:
:new: Add Simple type components

steelbrain committed:
:fire: Remove a few unnecessary public APIs

steelbrain committed:
:new: Allow disposal of loaded presets and components in pundle

steelbrain committed:
:new: Handle async watcherConfig.compile callbacks correctly

steelbrain committed:
:new: Add server sided logic of middleware

steelbrain committed:
:new: Add browser side of dev middleware

steelbrain committed:
:bug: Minor bug fixes in dev server

steelbrain committed:
:minibus: Correct dir name of reporter-cli

steelbrain committed:
:bug: Show correct font color in reporter-cli

steelbrain committed:
:bug: Minor bug fixes everywhere

steelbrain committed:
:new: Properly resolve multi-args in loader-js

steelbrain committed:
:new: Allow creating an express server in pundle-dev

steelbrain committed:
:no_entry: Temporarily disable CI on this branch

steelbrain committed:
:new: Add module.hot in HMR wrapper

steelbrain committed:
:new: Add rest of HMR

steelbrain committed:
Merge branch 'master' into steelbrain/v2

steelbrain committed:
:arrow_up: Upgrade typescript to v2 APIs

steelbrain committed:
:new: Add cli package

steelbrain committed:
:bug: Fix a typo

appsforartists commented on packages/transformer-typescript/README.md:
```diff
@@ -0,0 +1,7 @@
+# `pundle-transformer-babel`
```
"babel" :trollface:

steelbrain replied:
Copy/pasting has side-effects ha!

appsforartists commented on packages/transformer-typescript/src/index.js:
```diff
@@ -0,0 +1,35 @@
+  try {
+    typescriptPath = await this.resolve('typescript')
+  } catch (_) {
+    throw new MessageIssue('Unable to find babel-core in project root', 'error')
```
More "babel"; more :trollface:

appsforartists replied:
Why are you dynamically detecting typescript and throwing a custom error, rather than letting Node handle it?

steelbrain replied:
@appsforartists We are dynamically detecting typescript so if there's a new typescript release that has bugfixes, so we don't have to release a new version of this package. Requiring this from project allows us to require the project version of it

appsforartists replied:
Fair. peerDependency might be a better solution though.

steelbrain committed:
:bug: Fix a few typos
Thanks @appsforartists

appsforartists commented:
Looks like v2 is getting super close!

- Will the config support plugins/loaders?  Would be nice to dedupe that logic across my `devServer.js` and `build.js`.
- Any idea when an alpha will hit npm?

steelbrain commented:
@appsforartists We can push an alpha version as soon as we want to
and Yes. Each plugin/loader is gonna be a component. and a `Pundleconfig.js` can have `components: [...]` and `presets: [...]` so yes :)

steelbrain committed:
:new: Set NODE_ENV in CLI properly

steelbrain commented:
As another side note, you do not have to have different config files for dev/prod config. You just have to do `pundle -d` instead of `pundle` and it'll be a dev server

steelbrain committed:
v2.0.0-alpha1

steelbrain commented:
Another update, `Pundle@2.0.0-alpha1` is out on NPM 🎉

steelbrain committed:
:bug: Fix name in a package

steelbrain committed:
:bug: Fix a race condition in logging in npm-installer

steelbrain committed:
:bug: Fix a few bugs in progressed lock api callback

steelbrain committed:
:new: More bugfixes everywhere

appsforartists commented:
Looks like the README at https://github.com/motion/pundle/tree/steelbrain/v2/packages/pundle is still old.  Any docs anywhere for how v2 works/how to set it up/how it's different from v1?

steelbrain commented:
READMEs are a WIP. You'll have to use the flow types as docs for now. Also this Motion [Upgrade to v2](https://github.com/motion/motion/pull/410) PR should give you an idea of the changes in API

appsforartists commented:
Thanks.  FWIW, `Pundleconfig.js` is a weird name.  `pundle.config.js` maybe?

The initial cap and lack of word separation hurt my eyes.

steelbrain commented:
`pundle.config.js` is too long, I'd be down for `pundle.js`

appsforartists commented:
ha - it's one extra character 😝 

the subtle difference between `require('pundle')` and `require('./pundle')` makes me hesitant to truncate to `pundle`.  thoughts?

appsforartists commented:
whoa - i'm a contributor?!

steelbrain commented:
You did contribute the Typescript package 😉 
and Yeah I agree about the confusion, open to name suggestions :)

Also to keep the discussion on this PR focused, I've opened [an issue](https://github.com/motion/pundle/issues/85) about name of config file

steelbrain committed:
:bug: Fix an aggressive replacement bug in loader-js

appsforartists committed:
:bug: check for sourceMap before parsing it

steelbrain committed:
:new: Start listening early in http server instead of waiting for initial buildup

steelbrain committed:
Merge pull request #87 from appsforartists/p2ts

:bug: check for sourceMap before parsing it

appsforartists committed:
:bug: fix HMR in Safari

I get this error in Pundle@1 when I try to use HMR in Safari:

    SyntaxError: Unexpected keyword 'const'. Const declarations are not supported in strict mode.

Looks like this is the equivalent file in pundle@2, so I'm pre-emptively fixing it. :stuck_out_tongue_closed_eyes:

steelbrain committed:
Merge pull request #88 from appsforartists/safari-hmr

:bug: fix HMR in Safari

steelbrain committed:
:new: Add extension support to pundle-API and js/json loaders

steelbrain committed:
:new: Handle gracefully when Pundle config file doesn't exist in CLI

steelbrain committed:
:new: Process all known extensions in resolver by default

steelbrain committed:
:bug: Fix a bug in pundle-api about ext processing

steelbrain committed:
:bug: Fix race-conditions inside dev server's watcher and server

steelbrain committed:
:bug: Just another race condition fix in dev server

steelbrain committed:
:bug: Workaround a babel object-shorthand bug

steelbrain committed:
:bug: Fix a bug in imports merger
Because we are now tracking each one by a unique ID, merging uniques makes the app unusable because some of them are unresolvable (deleted by deduper)

steelbrain committed:
v2.0.0-alpha2

steelbrain commented:
Just published Pundle@v2.0.0-alpha2
It can now run our company app (several hundred deps and 15mb-ish) without a problem! 🎉

steelbrain committed:
:new: Allow redirect 404s to index
Fixes #89

steelbrain committed:
:new: Support passing notFoundToIndex to server directly

steelbrain committed:
:minibus: notFoundToIndex -> redirectNotFoundToIndex

steelbrain committed:
:new: Add pundle-loader-url package

steelbrain committed:
:fire: Remove publicPath from config for being useless

steelbrain committed:
v2.0.0-alpha3

steelbrain committed:
:bug: Fixes for HMR

steelbrain committed:
:racehorse: Push HMR before recompiling bundle

steelbrain committed:
:minibus: Use .pundle.js as the default name for config file
Fixes #85

steelbrain committed:
:bug: Fix a typo in pundle-browser doc

appsforartists committed:
:bug: Make pundle-cli accept command line args for devDirectory and port

steelbrain committed:
Merge pull request #94 from appsforartists/fix-cli

:bug: Make pundle-cli accept command line args for devDirectory and port

steelbrain committed:
:new: Provide base to parseInt

steelbrain committed:
:memo: Document the output and server config in types

steelbrain committed:
:new: Expose all of dev server options in Pundle config file

steelbrain committed:
:new: Add a new .pundle.js example

steelbrain committed:
:bug: Fix typo and remove a dep

appsforartists commented on packages/cli/src/index.js:
```diff
@@ -40,21 +40,23 @@ Pundle.create({
     return createServer(pundle, {
-      port: parseInt(program.port, 10),
+      port: program.port || config.server.port,
       directory: program.devDirectory,
```
This ought to also be settable in `config`.

The reason I opened #92 instead of filing a PR is that I didn't know how you wanted to check if `commander` had used the default, and you don't yet have a config to default to.  Maybe you delete the default from commander and change this to `config.server.directory || process.cwd()`.

steelbrain replied:
Ohhh, I actually made everything but this option configurable through the config file. (I've also added a copy-and-run config file that I'll be adding to `pundle init` command soon).

I'll make this configurable as well

appsforartists commented on packages/pundle/examples/Full Example (.pundle.js).js:
```diff
@@ -0,0 +1,22 @@
+    [require.resolve('pundle-preset-default'), {
+      resolver: { extensions: ['', '.js', '.json', '.ts', '.tsx'] },
+    }],
+  ],
```
YES!  So glad to have an example.

You should add the TypeScript plugin a) to show where plugins go and b) because you used `.tsx` in the example, but don't have a way to interpret those files.

steelbrain replied:
I am gonna add `pundle-preset-typescript`, so you don't have to do any of this configuration but just include the typescript preset along with the default one

appsforartists commented:
You should make that fix for `server.directory`, then update the alpha on NPM.  😃

steelbrain committed:
:new: Add pundle init command with sb-command

appsforartists commented:
🎉

steelbrain committed:
:bug: Properly throw resolution errors instead of config.entry

steelbrain committed:
:new: Allow configuring devDirectory through Pundle config file

steelbrain committed:
:bug: Fix prioritization bug in middleware
where it would serve bundle from fs instead of using the memory version if one existed on fs

steelbrain committed:
:bug: Fix mime in loader-url

steelbrain committed:
:new: Make HMR use websockets

steelbrain committed:
v2.0.0-alpha4

steelbrain committed:
:art: Change default HMR path

steelbrain commented:
Pundle v2.0.0-alpha4 is out! 🎉

steelbrain committed:
:arrow_up: Bump sb-command to v1

steelbrain committed:
:bug: Fix invalid determination of all known extensions

steelbrain committed:
:bug: Make resolver-default config.extensions consistent

steelbrain committed:
:new: Add pundle-preset-typescript

steelbrain committed:
:new: Make babel transformer accept exts instead of paths like typescript

steelbrain committed:
:new: Add Docs for preset-default

steelbrain committed:
:bug: Fix a bug in loader-js breaking import statements

steelbrain committed:
:bug: Copy old parents to new module in HMR

steelbrain committed:
:memo: Document how to exclude certain components in a preset

steelbrain committed:
:bug: Hide Sending HMR to 0 clients message

steelbrain committed:
:art: Reduce hmr retry time to 2 seconds

steelbrain committed:
v2.0.0-beta1

steelbrain committed:
:new: Allow configuring the log method in pundle-reporter-cli

steelbrain committed:
:bug: Fix config.redirectNotFoundToIndex support

steelbrain committed:
:new: Add debug for watcher tick

steelbrain committed:
:bug: Fix a bug where initial HMR would be huge

steelbrain committed:
:new: Add keep alive pings to HMR websocket

steelbrain committed:
:new: Prettyify FileIssues in transformer-babel

steelbrain committed:
:arrow_up: Bump to v2b2

steelbrain commented:
Pundle v2beta2 is out people! Should be stable enough to run an app properly
If you run into any bugs, please do let me know

steelbrain committed:
:bug: Support batch updates in file watcher

steelbrain committed:
:bug: Gracefully recover HMR state in case of module error

steelbrain committed:
:new: Use uws instead of ws as websocket server

steelbrain committed:
:new: Use more config options in default config file

steelbrain committed:
:bug: Fix detection of dev directory
Fixes #97

steelbrain committed:
:new: Minor format change in default config file

steelbrain committed:
:new: Add hmrHost to middleware config

steelbrain committed:
:new: Allow specifying hmrHost from config file

steelbrain committed:
:bug: Fix a typo

steelbrain committed:
:new: Add hmrReports config

steelbrain committed:
:new: Pipe reports into HMR sockets

steelbrain committed:
:new: Prepend the reports to body

steelbrain committed:
:new: Add Setup for development instructions

steelbrain committed:
:new: Set HMR reports off by default

steelbrain committed:
:arrow_up: Bump version to beta3

steelbrain committed:
:new: Split hmr deps into separate files

steelbrain commented:
We are getting there people!
<img src="https://cloud.githubusercontent.com/assets/4278113/21465257/1092876a-c950-11e6-8684-c14dabb7e8e6.png">

steelbrain committed:
:new: Add error reporting in browser with styles and everything

steelbrain committed:
:new: Send list of new files to HMR client

steelbrain committed:
:bug: Fix a bug where HMR would crash when new files were added

steelbrain committed:
:new: Recognize jsx as a default extension

steelbrain committed:
:new: Give up on HMR retries if connection refused

steelbrain committed:
:bug: Fix minor race condition in dev server

steelbrain committed:
:minibus: Rename config.server.devDirectory to config.server.rootDirectory for consistency

steelbrain committed:
:minibus: Rename --dev-directory to --server-root-directory in CLI

steelbrain committed:
:art: Internal changes across packages to match new name for dev root dir

steelbrain committed:
:bug: Fix a bug in Pundle where configuration file watcher lifecycle callbacks won't be invoked

steelbrain committed:
:new: Add transformer-babel to basic config example

steelbrain committed:
:new: Add full config example to 'pundle init full'

steelbrain committed:
:bug: Fix a typo

steelbrain committed:
Merge branch 'master' into steelbrain/v2

steelbrain commented:
Going to merge this as is

Gonna add docs and flow + eslint + typescript error support in a separate PR to make review easier
